Feature: Restrict Site Editor Access

efuller opened the pull request:
Fixes #4 

- Adds support for menus and customizer
- Restrict access to the site editor when on staging or production

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Implemented theme functions and definitions for WordPress theme customization.
  - Introduced namespace for organized code structure.
  - Custom site editor restrictions for non-local and non-development environments.

- **Documentation**
  - Updated theme metadata in `style.css`.

- **Refactor**
  - Enhanced theme setup process with additional support for menus and customizer features.

- **Style**
  - Applied new styles and structural comments for clarity in templates.

- **Bug Fixes**
  - Fixed accessibility to the site editor based on environment conditions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

efuller committed:
Add theme definition and required index template

efuller committed:
Add in support for the theme customizer

efuller committed:
Create, not Custom

efuller committed:
Consolidate to a theme setup

efuller committed:
Rename function

efuller committed:
Restrict access to site editor

efuller committed:
Bail if Caper is not available

efuller commented on inc/theme.php:
```diff
@@ -0,0 +1,25 @@
+	/**
+	 * Enable theme customizer.
+	 */
+	add_action( 'customize_register', '__return_true' );
```
This does seem to be an appropriate solution per https://developer.wordpress.org/reference/hooks/customize_register/#comment-6395

dlh01 replied:
That comment on wordpress.org is accurate, as far as it goes, but I think the bot is more correct. This isn't really the correct way to use `customize_register`, not least because actions don't return values. 

Taking a step back, why does the Customizer need to be enabled in this theme?

efuller replied:
@dlh01 I think the reasoning behind having the customizer in place is that it allows for things like setting a Site Icon, etc. @kevinfodness would you like to weigh in here? If we want to keep it activated, how would you recommend we do that?

kevinfodness replied:
Correct - WP has moved certain settings to full-site editing (e.g., you set the site icon using the site icon block in FSE rather than the customizer) but when FSE is disabled and the customizer isn't explicitly _enabled_, there is no way to edit things like the site icon and a few other settings that are tucked into the customizer (that I know of, at least, and I'd be happy to be proven wrong here).

efuller committed:
Rename constant

efuller committed:
Adjust function docblock

dlh01 commented on inc/site-editor.php:
```diff
@@ -0,0 +1,43 @@
+		// Prevent direct access to the Site Editor.
+		if ( 'site-editor.php' === $pagenow && ! wp_doing_ajax() ) {
+			Caper::deny_to_all()->primitives( 'edit_theme_options' );
+			Caper::grant_to( 'administrator' )->primitives( 'create_wordpress_theme_edit_menus' );
```
This capability doesn't appear to be used anywhere in this PR.

efuller replied:
I've removed!

dlh01 reviewed (commented):
The approach in this PR is inverted, almost, from the one I'm familiar with that we've taken on client projects. 

Here, users retain the capability to access the site editor until they actually try to access it. On the projects I know of, users are denied the capability to access the site editor and are instead granted the capability to edit menus "just in time":

- https://l.alley.dev/9bc7d5fe0b
- https://l.alley.dev/501b2fea78

So while I can speak to the approach from those client projects, can you say more about the benefits of this approach?

efuller committed:
Remove custom cap

efuller commented:
> The approach in this PR is inverted, almost, from the one I'm familiar with that we've taken on client projects.
> 
> Here, users retain the capability to access the site editor until they actually try to access it. On the projects I know of, users are denied the capability to access the site editor and are instead granted the capability to edit menus "just in time":
> 
> * https://l.alley.dev/9bc7d5fe0b
> * https://l.alley.dev/501b2fea78
> 
> So while I can speak to the approach from those client projects, can you say more about the benefits of this approach?

@dlh01 Thank you for the review. This work is stretching my T, so I appreciate the feedback.

I started off using the code here https://l.alley.dev/501b2fea78 but our use case for this project is a bit different. The Appearance menu needs to stay intact and function as it does by default, we just need to remove the `Editor` menu on `non-staging/production` environments.

The main benefit for my approach was that it seemed to work 😂 -- But with your reasoning, I can see why we might want to change it up.

I would welcome and suggestions or hints! In the meantime, I'll see what I could do based on your feedback.

dlh01 commented:
Ah! You're right that I was conflating "restrict site editor access" with "restore nav menu access," but that isn't strictly what's called for here.

In _that_ case, my suggestion would be that all you'll need on non-development environments is:

```
Caper::deny_to_all()->primitives( 'edit_theme_options' );
```

Users need `edit_theme_options` to access the site editor. If they don't have it, then the "Editor" menu will be removed automatically. There won't be a need to check `$pagenow` or doing-Ajax, either.

efuller committed:
Rebuild the appearance menu with items that are allowed to be accessed

efuller committed:
Add in a proper text domain

efuller committed:
Properly escape current URL

efuller commented on inc/site-editor.php:
@dlh01 @kevinfodness I've taken a different approach here for restricting the 'Editor' menu. Any feedback would be appreciated!

kevinfodness commented on inc/site-editor.php:
```diff
@@ -0,0 +1,90 @@
+
+		Caper::deny_to_all()->primitives( 'edit_theme_options' );
+		// Create a custom capability for menu items that should be displayed.
+		Caper::grant_to( 'administrator' )->primitives( 'cwt_edit_menus' );
```
We should spell out `create_wordpress_theme` here so that find and replace will update it to the project-specific thing:
```suggestion
		Caper::grant_to( 'administrator' )->primitives( 'create_wordpress_theme_edit_menus' );
```

kevinfodness commented on inc/site-editor.php:
```diff
@@ -0,0 +1,90 @@
+		}
+
+		// Conditionally re-grant access to menus page.
+		if ( 'nav-menus.php' === $pagenow && ! wp_doing_ajax() ) {
```
Why the `! wp_doing_ajax()` checks?

kevinfodness commented on inc/site-editor.php:
```diff
@@ -0,0 +1,90 @@
+function action__after_setup_theme(): void {
+	global $pagenow;
+
+	if ( ! in_array( wp_get_environment_type(), [ 'local', 'development' ], true ) ) {
+		// Caper is being installed via Create WordPress Project. If it's not available, bail.
+		if ( ! class_exists( 'Alley\WP\Caper' ) ) {
+			return;
+		}
+
+		Caper::deny_to_all()->primitives( 'edit_theme_options' );
+		// Create a custom capability for menu items that should be displayed.
+		Caper::grant_to( 'administrator' )->primitives( 'cwt_edit_menus' );
+
+		// Conditionally re-grant access to the customizer.
+		if ( ( 'customize.php' === $pagenow || 'index.php' === $pagenow ) && ! wp_doing_ajax() ) {
+			Caper::grant_to( 'administrator' )->primitives( 'edit_theme_options' )->at_priority( 100 );
+		}
+
+		// Conditionally re-grant access to menus page.
+		if ( 'nav-menus.php' === $pagenow && ! wp_doing_ajax() ) {
+			Caper::grant_to( 'administrator' )->primitives( 'edit_theme_options' )->at_priority( 100 );
+		}
+
+		// Rebuild the appearance menu.
+		add_action( 'admin_menu', __NAMESPACE__ . '\action__admin_menu' );
+	}
```
The function definition and filter ability come from WordPress, the files for which are not included in this codebase, as they are a third party dependency.

kevinfodness commented on inc/site-editor.php:
```diff
@@ -0,0 +1,90 @@
+	add_menu_page(
+		__( 'Appearance', 'create-wordpress-theme' ),
+		__( 'Appearance', 'create-wordpress-theme' ),
+		'cwt_edit_menus',
```
Here and anywhere else in this PR - expand `cwt` to `create_wordpress_theme` so find and replace can do its thing

kevinfodness commented on inc/site-editor.php:
```diff
@@ -0,0 +1,90 @@
+	remove_submenu_page( 'themes.php', $customize_url );
+	remove_submenu_page( 'themes.php', 'nav-menus.php' );
+
+	// Rebuild the Appearance menu.
```
Rather than rebuilding the Appearance menu, is it possible to just filter out the site editor item? My rationale here is that if other items were added to the appearance menu by plugins etc that we'd lose them in the process.

kevinfodness replied:
If part of the issue here is that we need to use custom capability checks, that could maybe be accomplished by a filter that is called when the submenu page is added?

efuller committed:
Replace cwt with create_wordpress_theme

efuller committed:
Allow capabilities to be re-applied if doing ajax

efuller committed:
Refactor site editor access

- Instead of rebuilding the appearance menu, just remove the Editor menu
- Deny edit_theme_options capabilities to all and re-apply if not on site-editor.php

efuller committed:
Add in the Customize menu item to the Appearance menu

efuller committed:
Update docblock

efuller commented:
@kevinfodness Would love to get your 👀 on the most recent changes. For restricting site editor, I took the approach of denying all and then conditionally allowing if we _aren't_ trying to access the site editor. The Editor menu item has also been removed.

The approach for adding in Customizer has been changed as well. Instead of using `customize_register`, the menu item is such being added in. // cc @dlh01

kevinfodness commented on inc/site-editor.php:
```diff
@@ -0,0 +1,41 @@
+
+		Caper::deny_to_all()->primitives( 'edit_theme_options' );
+
+		if ( 'site-editor.php' !== $pagenow ) {
```
If we are blocking access to `site-editor.php` entirely then this condition will always be true.

I propose removing Caper from this entirely, and just removing the menu item and redirecting if someone tries to hit it directly. If we find that we need to introduce additional capabilities checks in the future, we can always add them back in.

kevinfodness commented on inc/site-editor.php:
```diff
@@ -0,0 +1,41 @@
+ * Remove the Editor menu item from the appearance menu.
+ */
+function action__admin_menu(): void {
+	remove_submenu_page( 'themes.php', 'site-editor.php' );
```
The conditional check for environment should be moved here so the hook above can be removed entirely.

kevinfodness commented on inc/theme.php:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Setup theme defaults and registers support for various WordPress features.
+ */
+function action__after_theme_setup(): void {
```
Recommend matching this to the name of the action (`action__after_setup_theme`)

kevinfodness reviewed (approved):
One last tweak, then I think we're good 🍣

efuller committed:
Just remove the menu item and redirect if trying to access site-editor.php

efuller committed:
Fix action function name
